playbooks/robusta_playbooks/oom_killer.py

```diff
@@ -37,43 +55,211 @@ def oom_killer_enricher(event: NodeEvent):
+            resources = containers_spec_by_name[c_status.name].resources if c_status.name in containers_spec_by_name else None
+            memory_specs = self.get_memory_specs(resources)
+            oom_kill = OomKill(time=dt, pod_name=pod.metadata.name, container_name=c_status.name,
+                               image=c_status.image, memory_specs=memory_specs, reason=None)
+            oom_kill.reason = self.oom_kill_reason_investigator.get_reason(oom_kill)
+            oom_kills.append(oom_kill)
+
+        return oom_kills
+
+    def get_oom_killed_state(self, c_status: ContainerStatus) -> Optional[ContainerState]:
+        # Check if the container OOMKilled by inspecting the state field
+        if self.is_state_in_oom_status(c_status):
+            return c_status.state
+
+        # Check if the container OOMKilled by inspecting the lastState field
+        if self.is_last_state_in_oom_status(c_status):
+            return c_status.state
+
+        # OOMKilled state not found
+        return None
+
+    @staticmethod
+    def is_state_in_oom_status(status: ContainerStatus):
+        if not status.state:
+            return False
+        if not status.state.terminated:
+            return False
+        return status.state.terminated.reason == "OOMKilled"
+
+    @staticmethod
+    def is_last_state_in_oom_status(status: ContainerStatus):
+        if not status.lastState:
+            return False
+        if not status.lastState.terminated:
+            return False
+        return status.lastState.terminated.reason == "OOMKilled"
+
+    @staticmethod
+    def get_memory_specs(resources: Optional[ResourceRequirements]) -> MemorySpecs:
+        if resources is None:
+            return MemorySpecs()
+
+        mem_specs = MemorySpecs()
+
+        if resources.requests is not None and "memory" in resources.requests:
+            mem_specs.requests = resources.requests["memory"]
+
+        if resources.limits is not None and "memory" in resources.limits:
+            mem_specs.limits =  resources.limits["memory"]
+
+        return mem_specs
+
+
+class KubernetesOomKillReasonInvestigator(OomKillReasonInvestigator):
+    def __init__(self, node: Node, alert: PrometheusAlert, params: OomKillerEnricherParams):
+        self.config = params
+        self.memory_analyzer = MemoryAnalyzer(params.prometheus_url, alert.startsAt.tzinfo)
+        self.memory_transformer = K8sMemoryTransformer()
+        self.node = node
+        self.node_reason_calculated = False
+        self.node_reason = None
+
+    def get_reason(self, oom_kill: OomKill) -> str:
+        container_reason = self.get_busy_container_reason(oom_kill)
+        if container_reason is not None:
+            return container_reason
+
+        # Calculate if the node is the reason for the the OOMKill only once, rather than for every OomKill
+        if not self.node_reason_calculated:
+            self.node_reason = self.get_busy_node_reason()
```

Comment:
Can there be oom kill on the node, even with low node cpu utilization?
Like, lets say our node has 4gb of memory.
We have 3 pods with 1gb of memory request. Their actual usage is only 100MB each.

 Now, lets assume a process on the node, is trying to use 1.5GB
 Will it be oom killed (cause the node has 4gb, and 3gb is reserved for the pods, even though not used)
 If so, expect the cpu utilization here to be low. 
Can this happen?

Reply:
Did you mean low node **memory** utilization?

In any case, I don't think the scenario you described is possible.

Basically, the Scheduler splits the pods to 3 QoS classes: Guaranteed, Burstable and BestEffort.
-Guaranteed means all containers of the pod have a "request" and a "limit" in their spec, and that these are equal.
-Burstable means at least one container has a "request" in its spec, and yet it's not Guaranteed.
-BestEffort means there are no "request" in any container specs.
Now, when the Scheduler evicts pods, it will favor the Guaranteed class and will evict Burstable and BestEffort pods first (which means that Guaranteed pods will never be evicted due to a resource usage of other non-Guaranteed pods).

So in the case above, let's assume for the sake of the argument the pod whose process trying to use 1.5GB is BestEffort- which are the least prioritized, and that the 3 others are Guaranteed (request=limit).  In that case the scheduler will allow the BestEffort pod to run, as the node has enough resources for it. However, if the guaranteed pods will actually use 1GB each in some point in the future, the node will lack resources, and the scheduler will evict the BestEffort (1.5GB pod) first.

Reply:
Are you sure that's the case? 
Because only this week we had an issue with one of our customers, which had the exact same scenario.
A host process was oom killed, even though the host memory utilization was low.
It had few pods with memory requests, that weren't actually using most of it.